### DISCUSSION TITLE
Provide structured output for environment/interpreter discovery

### DISCUSSION THREAD

**Discussion by brettcannon** [brettcannon/python-launcher#168] [Ideas]
```JSONC
{
    // A unique identifier for the interpreter, i.e. the interpreter path.
    // Having it be the path to the interpreter also facilitates finding `build-details.json`.
    "path": "/home/brett/my-venvs/my-venv/bin/python",
    // An array specifying what is required to execute the interpreter.
    // The expectation is to append args to the end of the array before execution.
    // E.g. for conda environments:
    // ```
    // ["/path/to/conda", "run", "--path",
    //  "/home/brett/.conda/envs/conda-env", "--no-capture-output"]
    // ```
    "run": ["/home/brett/my-venvs/my-venv/bin/python"],
    "environment": { // Optional
        "type": "virtual",  // What type of environment, e.g. "virtual", "conda", etc.
        "name": "my-venv"  // Optional
    },
    // Is the result specific to the workspace?
    "contextual": true,
    // Who created this result/environment.
    "tool": "Python Launcher"
}
```

The results should be returned as [JSONL](https://jsonlines.org) so they can be streamed and processed immediately.

**Comment by brettcannon** [comment]
Feedback has been provided via https://discuss.python.org/t/what-information-do-you-gather-for-interpreters-and-environments/13468/ from folks involved in editors/IDEs.

**Comment by brettcannon** [comment]
Asked for feedback on Twitter via https://twitter.com/brettsky/status/1491168621900402689?s=20&t=v2lxc5_E1U0zLOnKOkjGEw .

**Comment by brettcannon** [comment]
https://twitter.com/brettsky/status/1491908748129955840 was a poll for potential binary prefix names. [Jannis suggested using `_` as prefix](https://twitter.com/jezdez/status/1491917980132945921) which does help with the tab completion conundrum, although I don't know what people would think of non-shell stuff having that prefix.

**Reply by brettcannon** [reply]
I am now leaning towards `_py_launcher_*`.

**Reply by jezdez** [reply]
I like this!

**Comment by brettcannon** [comment]
I'm not loving the `exact_match` key name. It's meant to represent that the user performed some action to suggest that the result is what they want when no search restriction is specified. Maybe `"context_sensitive"`?

**Reply by brettcannon** [reply]
Maybe `"halt"` as the name? Or `"best_fit"`? The thing I'm struggling with is wanting something to work in the `py` case, but then for it to not somehow mess up all the other cases (e.g. `--list`).

I also don't know if it's best to have a plug-in halt searching under the `py` case once a best-fitting to save on time by short-circuiting the search, or to let it still continue so that the tool calling the plug-in can decide whether that's good enough or whether it still wants to get a complete list of environments? In the latter case we could probably do away with arguments (or at least `--list`).

To be absolutely clear about the scenario I'm thinking of, consider a virtual environment in a `.venv` directory (or even `$VIRTUAL_ENV` being set for an activated environment). That's probably what the user wants as it's context-sensitive and required a user action to occur. And so searching for something better when running the command `py` is a waste of time. But how to communicate that? And taking the perspective of editors/IDEs, while knowing the best environment is great for automatic environment selection, you very well may want to still know all environments in case the user wants to change things. That would make having to ask the plug-ins to search again a bit of a waste and a time sink.

**Reply by brettcannon** [reply]
`"user_specified"`? `"user_selected"`?

**Comment by brettcannon** [comment]
Perhaps it would be handy to have an "activate" key that specifies the command to run in the shell to activate an environment?

**Comment by brettcannon** [comment]
We could have a `"watch"` key with a list of glob patterns. Then editors would know what to watch on disk to know when to call the locator again for updated information.

**Reply by brettcannon** [reply]
Not sure if there should be restrictions on the format, e.g. always POSIX-style?

**Reply by brettcannon** [reply]
[`GlobPattern` from VS Cod](https://code.visualstudio.com/api/references/vscode-api#GlobPattern)e doesn't support backslashes, but [`RelativePattern`](https://code.visualstudio.com/api/references/vscode-api#RelativePattern) will handle the conversion.

**Reply by brettcannon** [reply]
The [`glob` module](https://docs.python.org/3/library/glob.html#module-glob) will accept either format for paths.

**Reply by brettcannon** [reply]
I think I'm going to leave this out at least for v1.

**Comment by brettcannon** [comment]
My current thinking on CLI flags:

1. `--version`: Any potential hint on what to (not) search for, e.g. `==3.*` for `-3` for the Python Launcher. Search binaries MAY support this. Could also use `*`, `3`, and `3.10` as values.

A few open questions:
1. Should search binaries define `-h`/`--help`?
2. Should search binaries be told to never error out in the face of unknown options?
3. Is specifying the version restrictions (if available) even worth it?

**Reply by brettcannon** [reply]
The key thing here is how to service the two key use-cases:

1. The Python Launcher that wants the best answer the fastest it can get it.
2. Editors that want to know everything **and** what the suggested best answer is (which also services the Launcher case of needing to decide what to use).

It might come down to leaving the "best" selection logic to the caller and instead just provide ways to let the plug-ins know what they are searching for in order to maximize performance (e.g. `3` or `3.10`).

**Reply by brettcannon** [reply]
The other option for the second case is two calls: the best answer and then a potential second call to get the complete list. In VS Code's case, all you want is the best available to start, and then want the complete list only if users want to change their choice.

And the way to specify _any_ version is `>=0`, but it may be better to simply not specify a version restriction to signify that case and have `--list` be a thing. That would effectively make it [map to](https://github.com/brettcannon/python-launcher/blob/f744b4cd0271b88d55e3958f9cf33a2ce6eba4ce/src/lib.rs#L118-L125):

- `Any`: no options
- `MajorOnly`/`Exact`: `--version` depending on the argument (and there would be an expectation/standard to support either an e.g. `==3.10.*` format or `~=3.10.0` to say you want any 3.10 version for ease of parsing)
- List: `--list`

**Reply by brettcannon** [reply]
It might also make sense to either support or always pass in a `--workspace` flag that specifies where the search should be considered to be anchored at. This could also be the current working directory, but explicit is better than implicit. 😁

**Reply by brettcannon** [reply]
At the moment I'm leaning towards not worrying about any version argument/option, and only really thinking about whether to explicitly pass in the workspace/directory.

**Comment by brettcannon** [comment]
I wonder if the structured result here is the best solution, or a more flat one? My thinking is the nested ones make the entire object optional, but if you do define the object at all then you can have multiple required fields (which applies to 3/5 sub-objects). Being flat might make deserializing easier in certain languages.

Chances are I will keep the nested structure so in code it's easier to build up the details via various functions and then plug them together into the final object instead of having to unpack everything into a big, flat object at the end. Plus I find it mentally easier to understand where the information is expected to come from.

**Comment by brettcannon** [comment]
`"python_version"` or `"compatibility"`?

**Reply by brettcannon** [reply]
Or maybe `"version_info"` to keep with a pattern of matching where to get the info from `sys`?

**Comment by brettcannon** [comment]
Instead of a prefix and relying on `PATH` for locating locators exclusively, an environment variable could be supported to list more directories to prepend/extend `PATH` with.

**Comment by brettcannon** [comment]
The `"plugin"` name might not be the best name. Maybe `"finder"`? Totally a bikeshed magnet, but I don't know what the best term to use is.

**Reply by karrtikr** [reply]
The way I'm thinking It could also include pyenv, or a custom global python, etc. We're interested in the tool/plugin/app which created this environment/interpreter, so maybe `environment_source` could work.

```typescript
export enum EnvSource {
    Conda = 'Conda',
    Pipenv = 'PipEnv',
    Poetry = 'Poetry',
    VirtualEnv = 'VirtualEnv',
    Venv = 'Venv',
    VirtualEnvWrapper = 'VirtualEnvWrapper',
    WindowsStore = 'WindowsStore',
    Pyenv = 'Pyenv',
    Custom = 'Custom',
}
```

To me `finder` indicates how we found it (so `PATH`, 'WindowsRegistry` etc.) instead of what created it.

**Reply by brettcannon** [reply]
The tricky thing is it might not even have anything to do with what created it. Think of Graham's desire to detect some custom file that his shell reads when he enters a directory.

I might just drop the whole `"plug-in"` key anyway since it just helps to store where the info came from, but it isn't critical either as that can be stored external to the data without much issue.

**Comment by brettcannon** [comment]
Not critical to this proposal, but if a git-style plug-in system was created for the Launcher, then the found environment info could be passed down into it. For instance, let's say a `py-venv` plug-in was installed, then `py venv` could get passed in whatever the Launcher found, and then it could make the appropriate decision on how to create a virtual environment.

Granted, this doesn't save much compared to `py -m venv .venv`, but it does get the idea across.

**Reply by brettcannon** [reply]
This idea is tracked in https://github.com/brettcannon/python-launcher/discussions/222 .

**Comment by brettcannon** [comment]
It may be necessary to have a `shell` flag for executing if conda requires it (e.g. `conda.bat` and if Windows requires the shell be used for batch files).

**Comment by brettcannon** [comment]
It may not be necessary to flag that results are context-sensitive. That typically only comes up for environments, so if an environment wasn't returned then it . probably isn't critical. That means you can infer if a result is context-sensitive based on whether the results are environments.

This does fall apart, though, if you have interpreters where you have globally installed different packages.

But then perhaps the results should be flagged as context-aware, but a single result/key isn't necessary as the version selection from the Launcher can still perform its work.

**Comment by brettcannon** [comment]
I'm on the fence about including `interpreters.environment.activate`. There would need to be some key-consistency to work and people would need to know how to make the shell execute that string using e.g. `jq` to get the results out (or some `py activate` command). But it is something that most environments provide, so there is information to potentially share.

And this would purely be for non-Launcher users since the Launcher should implicitly be making activation pointless by doing the appropriate execution.

**Reply by brettcannon** [reply]
There's also an option to only collect the environment variables that would need to be set as part of the activation since that seems to be the key detail. That at least makes it potentially less shell-specific if things that are path-separated values like `PATH` are specified as an array. Otherwise one could assume the OS and not worry about the separators at all as that's really only a concern under potentially something like git-bash on Windows.

**Comment by brettcannon** [comment]
Nox feedback captured at https://github.com/wntrblm/nox/issues/622 .

**Comment by brettcannon** [comment]
Virtualenv feedback (which encompasses tox) captured at https://github.com/pypa/virtualenv/discussions/2345 .

**Comment by brettcannon** [comment]
Request for PDM feedback at https://github.com/pdm-project/pdm/discussions/1110 .

**Reply by brettcannon** [reply]
Should have a look at https://github.com/frostming/findpython .

**Comment by brettcannon** [comment]
When needing to execute using a shell, typically you want to pass in the string to use for that shell. That suggests that it may be better to not have a `shell` boolean but a string and an either-or scenario for specifying the command to run. That does make the command shell-specific for the situation of when the plug-in is run or having to list all possibilities for all shells. Using a string also complicates appending to the string as to whether arguments should go (and how to format them).

`conda` on Windows seems to be the specific case for needing shell support.

**Reply by brettcannon** [reply]
The conda team has suggested that requiring a shell for `conda run` can be viewed as a bug, so there isn't a need to support shells for execution.

**Comment by brettcannon** [comment]
Might need to make the Python version info optional due to conda environments potentially lacking a Python interpreter until the first `conda`-related command is run.

**Comment by brettcannon** [comment]
Specifying the bitness as an integer, separate from the CPU architecture, might be useful.

**Comment by brettcannon** [comment]
Providing a way to specify how to create an environment might be useful.

While Python provides `venv` for each and every interpreter installed, other tools may want to be able to provide a command that require input. This makes potentially structuring this information tricky as a finder/plug-in for e.g. Hatch will only have a single command which isn't tied to any Python interpreter while each individual Python interpreter on `PATH` will have its own way to create an environment. This also means that **all** Python info would be optional in the case where the tool doesn't know ahead of time what Python version will be used until it performs the environment creation or what command to use to use the environment.

It seems all the major environment management tools allow for some non-input command. For instance, `conda` can use an `environment.yml` to create an environment with no input from the user. Otherwise it seems most tools end up wanting a name and/or Python version.

- venv takes a location/name
- virtualenv takes a [location/name and optionally a version](https://virtualenv.pypa.io/en/latest/user_guide.html)
- conda can take [name and Python version](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#creating-an-environment-with-commands)
- Poetry can take [path or Python version](https://python-poetry.org/docs/managing-environments/#switching-between-environments)
- pdm can take [name and whatever back-end it ends up using](https://github.com/pdm-project/pdm/blob/main/pdm/cli/commands/venv/create.py) ([usage](https://pdm.fming.dev/latest/usage/venv/))
- pipenv can take the [Python version](https://docs.pipenv.org/basics/#specifying-versions-of-python)
- Hatch [doesn't take anything](https://hatch.pypa.io/latest/environment/) (on the command-line; [configured via `pyproject.toml`](https://hatch.pypa.io/latest/config/environment/#python-version))

This could also be a separate request to plug-ins/finders instead of part of finding out what's already available. But from a tool perspective, does it make sense to ask this question separately, or would folks want to know this all upfront? And which is the cheaper request: being told upfront about the command or asking later? And will it typically be specific to a finder/plug-in (e.g. "I chose `/usr/bin/python3.10`, and now I want an environment", or "what can I create an environment for?")?

The potential scenarios are:

- `py create-env`
- Editor selecting an interpreter and asking the user just before running code if they want to create an environment (which then potentially bleeds into installation post-environment creation)

**Reply by ofek** [reply]
To be specific: https://hatch.pypa.io/1.2/config/environment/#python-version

The default environment type is `virtual` which uses virtualenv but I expect all types to respect that option e.g. a Conda environment plugin.

**Reply by brettcannon** [reply]
Since the possibilities are pretty universal, having special support for something like `{python_version}` for Python version, `${name}` for the name, and `${location}` for the directory to place the environment in would work. And then any other `${}` variables would be up to some tool to prompt about.

Another question is whether it makes sense to support specifying the commands with varying amounts of detail. For instance, venv doesn't need a name; it's an option. So would being able to have an array of variations going from the bare minimum to whatever options are possible worth having? Or does just having the bare minimum make sense and if you need fancier then you are on your own?

**Reply by brettcannon** [reply]
This then also starts to bleed into whether providing package installation commands also makes sense? Since what is expected to specify installation requirements is standardized, the overall command to manage the installation can be just the overall command with what to install simply appended to the list at execution time.

**Reply by ofek** [reply]
Maybe that goes too far? I mean Hatch does support Docker envs with a plugin, and eventually k8s and other things, but standardizing seems like a setback for experimentation and choice in tools.

**Reply by brettcannon** [reply]
That's the trick with all of this, how far is enough to be handy and what's _too_ far 😅 .

I think there are 3 potential layers to this proposal:

1. Execution
2. Environment management
3. Installation management

Layer 1 is what sparked this whole thing and I think it still useful from a Python Launcher perspective. What I'm not sure about is layers 2 and 3 as that starts to get into potential UI requirements and perhaps that needs something higher-level than what a generic CLI like the Python Launcher could provide (e.g. requiring a custom extension for VS Code)?

@ofek do you think environment creation is generic enough and useful to work towards, and your only concern is package management?

**Reply by ofek** [reply]
I'm just not convinced 2 & 3 are worth it, only discovery is imo. Put another way, I think there's a reason why NodeJS, Ruby, etc. haven't standardized managing environments within containers or on other hosts.

However, I am aware Microsoft is the absolute best at dev tooling. So it's possible you all have some future vision of language deployment UX that would be super bad-ass and that this is simply a requirement to get things started. If so, then I suppose I'm amenable to the idea and will adapt Hatch to think in terms of this new spec 🙂

**Reply by brettcannon** [reply]
> I am aware Microsoft is the absolute best at dev tooling.

Thanks!

> it's possible you all have some future vision of language deployment UX that would be super bad-ass and that this is simply a requirement to get things started.

I wouldn't go that far. 😉 We are starting to think about how other VS Code extensions could participate in the creation/management of environments and installing/managing packages, and I was thinking out loud if this could somehow be made to be generic so other tools could benefit. I'm starting to think the answer is "no" and that locating and executing environments is where this is going to stop.

**Comment by brettcannon** [comment]
Request for feedback from Hatch at https://github.com/pypa/hatch/discussions/285 .

**Comment by brettcannon** [comment]
Request for feedback from pipenv at https://github.com/pypa/pipenv/discussions/5158 .

**Comment by brettcannon** [comment]
Request for feedback from poetry at https://github.com/python-poetry/poetry/discussions/5938 .

**Comment by brettcannon** [comment]
It has been proposed to include distributor details, but that's only available on Windows via the registry. I would rather get `sys` updated to expose this information and get it from there to make the data more generally available.

```cson
      "distributor": {  // Optional
          // PEP 514 (https://www.python.org/dev/peps/pep-0514/)
          "name": "Python Software Foundation",
          "url": "https://www.python.org"  // Optional
      },
```

**Reply by brettcannon** [reply]
Opened https://discuss.python.org/t/providing-sys-distributor-to-record-who-built-distributed-the-python-interpreter/18002/ to discuss this idea.

**Comment by Darsstar** [comment]
At a first glance this looks promising. I'll add this to my reading list for the next few weeks. I came here via the Poetry issue, I contributed some stuff, not a maintainer. But I also contributed to @pyenv-win, I was wondering if you have considered requesting feedback there and at @pyenv?

PS. I took a very short stab at adding Windows support to your launcher as my first serious Rust project. (To then later try to sneak in this kind of functionality for a better pyenv-win experience.) But quickly ran into the exit code dep being unix only and decided to try to improve other Python ecosystem parts first. It would be nice if this indeed ends up becoming a PEP, I would have to find a different Rust project to contribute to however 😉

**Reply by brettcannon** [reply]
> I was wondering if you have considered requesting feedback there and at @pyenv?

I have, just haven't gotten to it yet; I'm working my way through projects one at a time.

> I took a very short stab at adding Windows support to your launcher as my first serious Rust project. (To then later try to sneak in this kind of functionality for a better pyenv-win experience.) But quickly ran into the exit code dep being unix only

Thanks for at least taking a look! The code will need to be tweaked a bit to try and get Windows support, hence why I haven't tackled it quite yet 😅 (although I do plan to do it, so I won't put it off forever).

**Comment by brettcannon** [comment]
An option for the "I don't want to support shells but may need to thanks to `conda`", is to have `interpreters[].executable.run` be _either_ an array or a string, and if it's a string it requires the shell for execution.

**Reply by brettcannon** [reply]
I also still don't know how much environment variables may need to play into this. But if it does, it would probably be a dictionary of environment variable names and their values. Could probably map `null` to mean "delete", but otherwise assume that if e.g. `PATH` needed manipulation it would be provided already modified (e.g. all of this would be like `os.environ | env_dict` and that gets passed to `execv`).

**Reply by brettcannon** [reply]
The conda team has suggested that requiring a shell for `conda run` can be viewed as a bug, so there isn't a need to support shells.for execution.

**Comment by brettcannon** [comment]
Thinking about versioning the API in case a breaking change has to occur. Not sure if it should be at the top-level or nested in a `meta` table ala [PEP 691](https://peps.python.org/pep-0691/).

**Reply by brettcannon** [reply]
I think I would rather go with feature detection than versioning.

**Comment by brettcannon** [comment]
I'm wondering about recording the creator/manager of an environment? This could help identify when the environment may expect a specific tool to be used to manage it, e.g. Poetry would prefer you use it to (un)install packages. This would also indirectly record probably how the environment was discovered.

Another option is a boolean for `"managed"`, to flag something that isn't a plain virtual environment that you can simply `rm -rf`. This could all be combined into a "`managed_by"` that represents that an environment is managed and by what tool.

The impetus for this is a tool that uses the Launcher to find interpreters and environments. How is it to know if there are any restrictions in how it treats an environment? Can the tool install directly into it? What about deleting the environment? The reason this is at all tricky is you can have tools that want to be involved install an environment to `.venv`; Poetry and pipenv both allow for this. But how do you differentiate such a virtual environment from one created via `venv`? That might suggest the boolean is good enough to say, "leave the environment as-is," and then have some external mechanism to specify _how_ to manage the environment.

There's also the desire to potentially expose this to users in tools using the Launcher to find environments. For instance, users may want to know the environment was picked up from Poetry versus some random other environment. While showing the path to the environment would help clarify that, it doesn't mean the tool knows about the environment manager nor that the path is the shortest thing to share in a bit of UI.

**Reply by karrtikr** [reply]
I think `managed_by` is not something specific to just "environments", for eg. consider the pyenv global installs. So maybe we should pull it out of the `"environment"` umbrella.

**Reply by brettcannon** [reply]
I don't think whether an interpreter is installed by the OS or by pyenv is important. In this case I was adding the `"managed_by"` key for environments to signal to folks how to (un)install packages in that environment, etc. But once again that may be unnecessary in this instance and expand to a higher-level API where environment management is specified.

**Comment by brettcannon** [comment]
I'm contemplating changing `"interpreters"` to not be a map but a list with a `"path"` key. This would let locators have a concept of order in case the Launcher needs some help in choosing between interpreters (think multiple Python 3.10 interpreters and using the order found on `PATH` as the way to choose preference).

I have gone back and forth on this as I typically want the Launcher to decide what to use, but there is an inherent, preferential order to things just like `PATH`, and so not capturing that seems like throwing away information.

**Comment by brettcannon** [comment]
Someone suggested to me to support a `.py_launch.json` file which would statically contain details in the current (or parent) directory. That would avoid execution overhead in cases where such a file might make sense.

**Reply by brettcannon** [reply]
I don't think this is necessary for v1.

**Comment by brettcannon** [comment]
Right now the bitness and CPU architecture are two separate things: `executable.bits` and `executable.arch`. Perhaps it makes more sense to unify it into just `architecture` or `machine` to match `platform.machine()`/`platform_machine` [packaging marker](https://peps.python.org/pep-0508/#environment-markers)? Somewhat depends on whether the info is for the actual CPU or how Python was compiled.

**Reply by brettcannon** [reply]
I dropped CPU architecture because I can't find a reliable way to tell from Python what the binary is compiled for compared to the CPU itself (i.e. `platform.architecture()` is empty on macOS where you have x64 or Apple Silicon to contend with).

**Comment by brettcannon** [comment]
Request for feedback from pyenv at https://github.com/pyenv/pyenv/discussions/2472 .

/cc @Darsstar

**Comment by brettcannon** [comment]
Twitter poll at https://twitter.com/brettsky/status/1578101990625710080?s=20&t=JCU0NXB3sEwbBIs9qqT6_g about whether to prefer interpreters with more version details or less (e.g. select `3.10` or `3.10.2` when a `3.10` interpreter is requested).

**Reply by brettcannon** [reply]
And the winner was ... no one! 😅 Too close to call, so I plan to say "unknown" < "0".

**Comment by brettcannon** [comment]
I'm realizing that this will cause a shift in Launcher semantics. Right now if you specify a version constraint, checking `.venv` is skipped. But that may not be what people want if they specify a locator that supports multiple virtual environments (e.g. something for Nox environments to poke around as to why a test failed). So I think the Launcher will need to decide to either prefer the newest environment, or always prefer the newest whatever regardless of the environment.

**Reply by brettcannon** [reply]
Actually, if the discovery code is moved into a library, then the Launcher can continue to prefer `.venv`. Effectively the Launcher and explicitly look for `.venv` and use it if found. Otherwise the Launcher can perform a full search and then that can be used to decide what to prefer. This keeps the "`.venv` beats all" semantics while still leveraging expanded results.

**Comment by brettcannon** [comment]
To record my thought process on something ...

I was considering dropping the `environment.type` key since it doesn't play into how you run an intepreter thanks to `executable.run`. But it **does** play into how to you choose to manually interact with it, e.g. if your editor tells you it's a conda environment and thus you are going to use `conda install` instead of `pip install`. So it's more of a UX hint than critical piece of information on how to run code.

**Comment by brettcannon** [comment]
One thing I'm thinking about is how to display something in a list of interpreters/environments (i.e. `py --list`)? I do know folks are asking for a `which`-like experience to just get at the path of the interpreter (see https://github.com/brettcannon/python-launcher/discussions/226). Now as it stands, `"path_id"` won't serve that purpose because of its "unambiguous but as weakly specified as possible" design. Now if that was changed to "as tightly specified as possible" (e.g. virtual environments would be expected to always specify the `pythonN.N` interpreter path), then that could play double-duty for both the listing and the UUID for what was found. But the key reason I did the initial "weak" design was in case someone found `~/.virtualenvs/my-venv/bin/python3.11` as just an interpreter as well as a virtual environment. By saying virtual environments would have the ID be `~/.virtualenvs/my-venv`, it would work around this ambiguity. But maybe it should be recognized there's always the chance of ambiguity and not worry about it? That way the consumer of the data is left with deciding what should take precedence?

Another solution is to introduce something like a `"path_display"` that could be ambiguous, but is purely for situations like `py --list`. But that might be a bit too redundant?

**Comment by brettcannon** [comment]
https://hynek.me/til/python-project-local-venvs/ covers how direnv users do things (from https://mastodon.social/@hynek/109561175719682337).

**Comment by dhellmann** [comment]
How dynamic is the data that would be retrieved? How often do you expect the information for a project to change? Would it make sense to keep the information in a file associated with the project, instead of invoking another program to look it up each time? For example, I could envision a plugin for virtualenvwrapper that updated a file (with a format defined by the launcher) with data each time an env is linked to a project directory via setvirtualenvproject or removed with rmvirtualenv. Then the launcher just needs to read that config file, and long running tools like IDEs could cache the results.

**Reply by brettcannon** [reply]
> How dynamic is the data that would be retrieved?

I've thought about potentially caching and such (to the point of letting locators specify glob patterns to file watch for changes; both are one of the hidden comments), but in the end I figured simpler is better and so I don't plan to directly support any caching. If a locator like one for virtualenvwrapper wanted to cache a file, it totally could and is welcome to do so, but I think having the Launcher itself make any assumptions about the freshness/staleness of data is asking for trouble.

I also assume that for most people, they probably going to have 0, 1, maybe 2 locators, and so the overhead of calling them is going to be rather small relatively speaking compared to probably what you're about to do with Python itself.


> How often do you expect the information for a project to change?

I have no idea, hence why I don't want to handle the caching. 😉



> Would it make sense to keep the information in a file associated with the project

This is why I have thought about whether https://github.com/brettcannon/python-launcher/discussions/165 makes sense to support directly and why I can't make up my mind about it.



> instead of invoking another program to look it up each time?

Based on how trying to cache things in VS Code for environments has gone, I would rather avoid that as much as possible and assume folks are not going to use locators that have a drastic cost to run.

**Reply by dhellmann** [reply]
> Based on how trying to cache things in VS Code for environments has gone, I would rather avoid that as much as possible and assume folks are not going to use locators that have a drastic cost to run.

OK, I can buy that running a program each time to answer the question is the best approach.

**Comment by brettcannon** [comment]
I brought back the `executable.architecture` field as I realized there wasn't going to be a way to tell when someone was using a WebAssembly build of Python vs a native binary that happened to also be 32-bits. Technically there's still a difference between something built for WASI or generic WebAssembly for the browser using Emscripten, but I don't expect much overlap in use cases in that regard. Plus if it ends up being that critical then people could probably abuse the `architecture` field somewhat to include that detail.

**Comment by brettcannon** [comment]
Having `path_id` be a path to a binary would be useful for [PEP 668: `EXTERNALLY-MANAGED`](https://peps.python.org/pep-0668/) if `    sysconfig.get_path("stdlib", sysconfig.get_default_scheme())` can be statically calculated, else `executable.run` solves the problem of getting that information.

**Comment by brettcannon** [comment]
I moved `run` to the top-level as I don't think any of this information is useful if you can't actually execute the interpreter.

**Comment by brettcannon** [comment]
Since this proposal started, https://packaging.python.org/en/latest/specifications/build-details/ has come into existence, so there's no need to duplicate that information here.
